cmd/pki.go

```diff
@@ -0,0 +1,354 @@
+func buildCACert(template *x509.Certificate, privateKey interface{}) ([]byte, *x509.Certificate) {
+	template.IsCA = true
+	template.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageCRLSign
+
+	derBytes := buildCert(template, template, privateKey, privateKey)
+
+	caCert, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		log.Fatalf("Failed to parse CA certificate: %s", err)
+	}
+
+	return derBytes, caCert
+}
+
+// Wrapper function to generate the server certificate
+func buildServerCert(template *x509.Certificate, host string, privateKey interface{}, signerCert *x509.Certificate, signerPrivKey interface{}) []byte {
+
+	hosts := strings.Split(host, ",")
+	for _, h := range hosts {
+		if ip := net.ParseIP(h); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, h)
+		}
+	}
+
+	template.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+
+	return buildCert(template, signerCert, privateKey, signerPrivKey)
+}
+
+// Wrapper function to generate the client certificate
+func buildClientCert(template *x509.Certificate, id string, privateKey interface{}, signerCert *x509.Certificate, signerPrivKey interface{}) []byte {
+	template.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
+	template.Subject.CommonName = id
+
+	return buildCert(template, signerCert, privateKey, signerPrivKey)
+}
+
+// Write key material to disk. Note : the CA's private key is not written to disk.
+func writeCertAndKeyToDisk(entityType string, privateKey interface{}, certFilename string, keyFilename string, derBytes []byte) {
+
+	// Make sure target directory exists
+	err := os.MkdirAll(filepath.Dir(certFilename), 0700)
+	if err != nil {
+		log.Fatalf("failed to create directory for cert : %s.\n%s", certFilename, err)
+	}
+
+	// Write Certificate to File
+	certOut, err := os.Create(certFilename)
+	if err != nil {
+		log.Fatalf("failed to open cert.pem for writing: %s", err)
+	}
+
+	pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	certOut.Close()
+	// log.Print("wrote " + certFilename + "\n")
+
+	// Write key to file if it is not a CA
+	if entityType != "CA" || keepPrivCA {
+		keyOut, err := os.OpenFile(keyFilename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+		if err != nil {
+			log.Print("failed to open "+keyFilename+"  for writing:", err)
+		}
+		pem.Encode(keyOut, pemBlockForKey(privateKey))
+		keyOut.Close()
+		// log.Print("wrote " + keyFilename + "\n")
+		// log.Print("Successfully achieved job for " + keyFilename + "\n")
+	}
+
+}
+
+// Creates the certificate for the given entity and writes key material to disk. In case of the CA, also return the certificate and private key to sign the end entity parties
+func create(entityID string, validFor time.Duration, entityType string, ecdsaCurve string, signerCert *x509.Certificate, signerPrivKey interface{}, organisation string, destDir string) (*x509.Certificate, interface{}, string, string) {
```

Comment:
Can you split this function ?
Line 337, you are passing 2 nil parameter and use only 2 return value
Line 343 and 348 should call a simpler dedicated function ?
If this is not a CA (line 318) you always return 2 nil value
`createServer`, `createClient` and `createCA` maybe ?

Reply:
You're right. I kept the structure of my POC. Splitting will clear the code.